f3_test.go

```diff
@@ -385,87 +422,59 @@ func (e *testEnv) advance() {
 	e.clock.Add(1 * time.Second)
 }
 
-func (e *testEnv) waitForManifestChange(prev *manifest.Manifest, timeout time.Duration) {
-	e.t.Helper()
-	e.waitForCondition(func() bool {
-		for _, n := range e.nodes {
-			if !n.f3.IsRunning() {
-				continue
-			}
-
-			m := n.f3.Manifest()
-			if m == nil {
-				return false
-			}
+func (e *testEnv) withNodes(n int) *testEnv {
+	for i := 0; i < n; i++ {
+		e.addNode()
+	}
 
-			if prev.Equal(m) {
-				return false
-			}
-		}
-		return true
-	}, timeout)
+	return e
 }
 
-func newTestEnvironment(t *testing.T, n int, dynamicManifest bool) *testEnv {
+func newTestEnvironment(t *testing.T) *testEnv {
 	ctx, cancel := context.WithCancel(context.Background())
 	ctx, clk := clock.WithMockClock(ctx)
 	grp, ctx := errgroup.WithContext(ctx)
-	env := &testEnv{t: t, errgrp: grp, testCtx: ctx, net: mocknet.New(), clock: clk, tempDir: t.TempDir()}
+	env := &testEnv{
+		t:              t,
+		errgrp:         grp,
+		testCtx:        ctx,
+		net:            mocknet.New(),
+		clock:          clk,
+		tempDir:        t.TempDir(),
+		manifest:       base,
+		signingBackend: signing.NewFakeBackend(),
+	}
 
 	// Cleanup on exit.
 	env.t.Cleanup(func() {
 		require.NoError(env.t, env.net.Close())
 
 		cancel()
 		for _, n := range env.nodes {
-			require.NoError(env.t, n.f3.Stop(context.Background()))
+			if n.f3 != nil {
+				require.NoError(env.t, n.f3.Stop(context.Background()))
+			}
 		}
 		env.clock.Add(500 * time.Second)
 		require.NoError(env.t, env.errgrp.Wait())
 	})
 
-	// populate manifest
-	m := base
-	initialPowerTable := gpbft.PowerEntries{}
-
-	env.signingBackend = signing.NewFakeBackend()
-	for i := 0; i < n; i++ {
-		pubkey, _ := env.signingBackend.GenerateKey()
-
-		initialPowerTable = append(initialPowerTable, gpbft.PowerEntry{
-			ID:     gpbft.ActorID(i),
-			PubKey: pubkey,
-			Power:  gpbft.NewStoragePower(1000),
-		})
-	}
-	env.manifest = m
-	env.ec = consensus.NewFakeEC(ctx, 1, m.BootstrapEpoch+m.EC.Finality, m.EC.Period, initialPowerTable)
-
-	var manifestServer peer.ID
-	if dynamicManifest {
-		env.newManifestSender()
-		manifestServer = env.manifestSender.SenderID()
-	}
-
-	// initialize nodes
-	for i := 0; i < n; i++ {
-		env.initNode(i, manifestServer)
-	}
-
 	return env
 }
 
-func (e *testEnv) initNode(i int, manifestServer peer.ID) {
-	n, err := e.newF3Instance(i, manifestServer)
+func (e *testEnv) addNode() *testNode {
```

Comment:
This creates a test node but doesn't actually construct the F3 instance. That happens when `init` and/or `start` is called, giving us a chance to mess with the manifest/ec/etc.